src/server/bundle.ts

```diff
@@ -98,58 +92,170 @@ export class Bundler {
+
+  const saveBundleOnKv = async () => {
+    const start = performance.now();
+
+    // We need to save chunks first, islands/plugins last so we address esm.sh build instabilities
+    const isChunk = /\/chunk-[a-zA-Z0-9]*.js/;
+    const chunksFirst = [...inMemoryCache.keys()].sort((a, b) => {
+      const aIsChunk = isChunk.test(a);
+      const bIsChunk = isChunk.test(b);
+      const cmp = a > b ? 1 : a < b ? -1 : 0;
+      return aIsChunk && bIsChunk ? cmp : aIsChunk ? -10 : bIsChunk ? 10 : cmp;
+    });
+
+    for (const file of chunksFirst) {
+      const content = inMemoryCache.get(file)!;
+      await kv?.saveFile(file, content).catch((e) =>
+        console.error(`Error: Saving file to KV failed ${file}\n`, e)
+      );
+    }
+
+    console.info(
+      `💾 Saving bundle to Deno.KV took ${
+        ((performance.now() - start) / 1e3).toFixed(2)
+      }s`,
+    );
+  };
+
+  const clearOldData = async () => {
+    const start = performance.now();
+    await kv?.housekeep();
+    console.info(
+      `🧹 Housekeeping stale data on Deno.KV took ${
+        ((performance.now() - start) / 1e3).toFixed(2)
+      }s`,
+    );
+  };
+
+  const bundleAndCacheOnce = once(async () => {
+    const build = await bundler();
 
-    const cache = new Map<string, Uint8Array>();
+    // Save Build on inMemoryCache
     const absDirUrlLength = toFileUrl(absWorkingDir).href.length;
-    for (const file of bundle.outputFiles) {
-      cache.set(
+    for (const file of build.outputFiles!) {
+      inMemoryCache.set(
         toFileUrl(file.path).href.substring(absDirUrlLength),
         file.contents,
       );
     }
-    cache.set(
+    inMemoryCache.set(
       "/metafile.json",
-      new TextEncoder().encode(JSON.stringify(bundle.metafile)),
+      new TextEncoder().encode(JSON.stringify(build.metafile)),
     );
-    this.#cache = cache;
 
-    return;
-  }
+    // Fire & forget saving on KV & housekeeping
+    kv && saveBundleOnKv().then(clearOldData).catch((e) => console.error(e));
+  });
 
-  async cache(): Promise<Map<string, Uint8Array>> {
-    if (this.#cache === undefined) {
-      this.#cache = this.bundle();
-    }
-    if (this.#cache instanceof Promise) {
-      await this.#cache;
-    }
-    return this.#cache as Map<string, Uint8Array>;
-  }
+  return {
+    get: async (path: string) => {
+      const file = inMemoryCache.get(path) ||
+        await kv?.getFile(path).catch(() => null);
 
-  async get(path: string): Promise<Uint8Array | null> {
-    const cache = await this.cache();
-    return cache.get(path) ?? null;
-  }
+      if (file instanceof ReadableStream) {
+        console.info(`🚤 streaming directly from Deno.KV: ${path}`);
+      }
 
-  // getPreloads(path: string): string[] {
-  //   return this.#preloads.get(path) ?? [];
-  // }
-}
+      if (file) return file;
+
+      // Cache not found on both Kv or inMemory, let's build and cache it
+      await bundleAndCacheOnce();
+
+      return inMemoryCache.get(path);
```

Comment:
Any change that `bundleAndCacheOnce` will not succeed in putting `path` in `inMemoryCache`?

Reply:
We currently have the same behavior, so I don't think this is an issue